include/swift/AST/Attr.h

```diff
@@ -1024,34 +1027,84 @@ class SynthesizedProtocolAttr : public DeclAttribute {
   }
 };
 
+class AbstractFunctionDecl;
+
 /// The @_specialize attribute, which forces specialization on the specified
 /// type list.
 class SpecializeAttr : public DeclAttribute {
-  unsigned numTypes;
-  ConcreteDeclRef specializedDecl;
+public:
+  enum class SpecializationKind {
+    Full,
+    Partial
+  };
 
-  TypeLoc *getTypeLocData() {
-    return reinterpret_cast<TypeLoc *>(this + 1);
+private:
+  AbstractFunctionDecl *F;
```

Comment:
I think you shouldn't store the parent function inside the attribute, and just pass it in when necessary - what do you think?